library/test/src/lib.rs

```diff
@@ -591,6 +593,28 @@ fn run_test_in_process(
     // Buffer for capturing standard I/O
     let data = Arc::new(Mutex::new(Vec::new()));
 
+    // If we encounter a non-unwinding panic, stop capturing output to ensure that the
+    // non-unwinding panic message is visible.
+    // We also acquire the locks for both output streams to prevent output from other threads
+    // from interleaving with the panic message or appearing after it.
+    let builtin_panic_hook = panic::take_hook();
+    let hook = Box::new({
+        let data = data.clone();
+        move |info: &'_ PanicInfo<'_>| {
+            if !info.can_unwind() {
+                io::set_output_capture(None);
+                std::mem::forget(crate::io::stderr().lock());
+                let mut stdout = ManuallyDrop::new(crate::io::stdout().lock());
+                if let Ok(data) = data.lock() {
+                    let _ = stdout.write_all(&data);
+                    let _ = stdout.flush();
+                }
+            }
+            builtin_panic_hook(info);
+        }
+    });
+    panic::set_hook(hook);
```

Comment:
Shouldn't the hook be reset to its original form after the test is done?
What does this do in a concurrent situation, it ends up with a chain of these hooks calling each other or so and printing the `data` from the first since all others will fail to get the lock?

Reply:
In case the capture is reset the process would abort anyway. As for running multiple tests, you're right. This set_hook should probably be moved to before the loop over all tests such that it only gets registered once. No need to reset it in that case as that function is normally only called once per process anyway.

Reply:
This was actually buggy for other reasons.

The hooks _do_ chain, and the whole chain does actually run because output capture is a thread-local and and the stdout/stderr locks are reentrant, the previous implementation prints the captured output from _every_ test, most recent test first :facepalm: